rfd/0138-postgres-backend.md

```diff
@@ -0,0 +1,105 @@
+
+Each auth server running the backend will periodically attempt to delete expired items, at an interval tentatively set to 30 seconds, deleting batches of expired items (in different transactions) until no more items can be deleted.
+
+To generate an event stream we'll connect to the database with a regular connection, create a temporary logical replication slot, then poll changes on it via the [`pg_logical_slot_get_changes`](https://www.postgresql.org/docs/current/functions-admin.html#FUNCTIONS-REPLICATION) function. As the builtin `pgoutput` logical decoding plugin cannot be used without errors via the SQL interface except in very recent versions of Postgres (11.20, 12.15, 13.11, 14.8 or 15.3, released in May 2023), we use the third-party but very popular [`wal2json`](https://github.com/eulerto/wal2json) plugin, which is available by default on the managed Postgres offerings in Azure, AWS and GCP, and is installable from the Debian/Ubuntu official packages or from the Postgres APT and YUM repos.
+
+The `pg_logical_slot_get_changes` function (with the appropriate `wal2json` options) will return one entry per change, which maps nicely to the backend event model; updates and inserts will be rendered as `OpPut` events, deletes as `OpDelete` events.
+
+### General architecture of the audit log
+
+The database schema for the audit log will be composed of just an `audit` table with the appropriate indexes for searching:
+
+```pgsql
+CREATE TABLE audit (
+	event_time timestamp NOT NULL,
+	audit_id uuid NOT NULL,
+	event_type text NOT NULL,
+	session_id uuid NOT NULL,
+	event_data json NOT NULL,
+	creation_time timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
+	PRIMARY KEY (event_time, audit_id)
+);
+CREATE INDEX audit_session_id_idx ON audit (session_id, event_time, audit_id) WHERE session_id != '00000000-0000-0000-0000-000000000000';
+CREATE INDEX audit_creation_time_idx ON audit (creation_time);
+```
+
+The database schema is made to be compatible with the one for the backend, although we are going to recommend storing the backend and the audit log in two different databases (having both on the same database cluster is going to be fine), because logical decoding is applied per-database, and keeping both in the same database will require the logical decoding plugin to do extra work to filter out the changes to the audit log.
+
+The `audit_id` column is a UUID chosen at random at insert time, unrelated to the event data, only used to disambiguate events for the purpose of pagination; `creation_time`, likewise, is only used for deleting events after some retention period (defaulting to one year).
+
+With minor modifications, this schema can support range based partitioning by `event_time`, creating partitions by hand or using `pg_partman`. The initial implementation will not use partitioning, but setting up the table layout for it will allow for a much easier migration if and when we end up supporting it.
```

Comment:
Open question: should we set up a partitioned table with a single partition anyway, or is that something that can be handled at a later time? It should be possible to convert the table into a partitioned table with a single partition relatively quickly in a transaction, so it should be doable without downtime, but it's surely easier to do as we're setting up the schema for the first time

Reply:
Migrating from a single table to a partitioned table by renaming the original table is plenty fast, we can do it at a later time (if ever).